Implement embedding extraction

palonso opened the pull request.

palonso committed:
Reimplement feature extractors [WIP]

The extractors are implemented in classes to facilitate initializating them only once at the
begging of the feature extraction. This is specially beneficial for the embedding extractors
based in TensorFlow.

The embedding extractors reuse the mel-spectrogram extractors when possible.

TODO: Port Splitter and TempoCNN

palonso committed:
Reimplement preprocess and preprocess_crosseval

preprocess.py is reimplemented to support the new feature extractors.
It is designed rely on the standard Essentia functions by default, and it only imports
TensorFlow-related algorithms when embeddings are required.

preprocess_crosseval.py is modify to rely mostly in preprocess.py methods. It's basically
a wrapper that works with arguments instead of a configuration file (which is useful for
external datasets). This affects where the script writes the index.tsv file: in `data_dir`,
 instead of `datadir/metadata` as before.

palonso committed:
Add pretrained models

Bomme commented on src/feature_melspectrogram.py:
```diff
@@ -0,0 +1,117 @@
+
+        self.loader.audio >> self.frameCutter.signal
+        self.frameCutter.frame >> self.mels.frame
+        self.mels.bands >> (self.pool, 'mel_bands')
```
`>>` 🤔

palonso replied:
This is the `connect` operator for [Essentia streaming](https://essentia.upf.edu/streaming_architecture.html#essentia-streaming-mode-architecture) algorithms

Bomme replied:
But is this valid python code? It looks like it was simply copied from C++!

palonso replied:
Yes, this is how Essentia is used in streaming mode in Python. 
Maybe [this](https://essentia.upf.edu/essentia_python_tutorial.html#using-essentia-in-streaming-mode) is the best resource to get started with it in case you want to learn more 🙂

palonso committed:
change name: effnet -> effnet_b0

palonso committed:
Remove global variables

palonso committed:
Make melspectrogram_to_batch private

palonso committed:
Discard parallelization options

Now all the parallelization happends through TensorFlow by setting the batch size

palonso committed:
Remove reimplemented OpenL3 feature

palonso committed:
Update embedding name: ol3 -> openl3
